Fix the damage particle PR

TUsama opened the pull request.

TUsama committed:
add particle and parckets

TUsama committed:
change name, and more implements

TUsama committed:
Create ExileEvent.java

TUsama committed:
Merge branch 'event' into dmg

TUsama committed:
add NotifyClientToSpawnParticleEvent

TUsama committed:
implement the particle pipeline

TUsama committed:
implement the damage particle, now it can actually work

TUsama committed:
format

TUsama committed:
code clean up

TUsama committed:
move scale change into strategy

TUsama committed:
implement HealParticle

TUsama committed:
Create Wynn.java

TUsama committed:
change name, move sound play logic into InteractionResultHandler, prob we can play different sound for dodge and resist

TUsama committed:
heal particle

TUsama committed:
implement row style

TUsama committed:
remove event-related code in this PR, this part should not in this PR

TUsama committed:
remove test mod

TUsama committed:
recover some old code in case we need to enable it again

TUsama committed:
clean up

TUsama committed:
Merge branch '1.20-Forge' into dmg

TUsama committed:
recover

TUsama committed:
move the material to other place to avoid invoke Particle.class on server

TUsama committed:
fix invoke client config on server, also rename the strategy

TUsama committed:
make the particle text actually always can be seen

TUsama committed:
use old config option to control the particle notifier

RobertSkalko commented on src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/InteractionResultHandler.java:
```diff
@@ -20,27 +20,43 @@
+        NULLIFIED_DAMAGE(IParticleSpawnMaterial.Type.DODGE) {
+            @Override
+            public Supplier<BiConsumer<IParticleSpawnMaterial, Entity>> getStrategy() {
+                return () -> ClientConfigs.getConfig().DAMAGE_PARTICLE_STYLE.get().nullifiedDamageStrategy;
```
Uh what
Just return DAMAGE_PARTICLE_STYLE and grab the varible needed elsewhere

RobertSkalko replied:
It doesn't need 3 overrides that do mostly the same thing, and it doesn't need a supplier if it's already a method

TUsama replied:
The ParticleSpawnType exists on both sides, but BiConsumer only exists on the client, which is why it caused a crash in the last build. 
The current code structure is that the server sends the particle material to the client. The material includes a ParticleSpawnType and can be obtained on the client side. From the ParticleSpawnType, client could know which consumer should be used for the material.

RobertSkalko replied:
I don't get it. Just return IParticleSpawnMaterial in the method and only call the method on the client?

RobertSkalko replied:
Also i generally try to never use biconsumers and stuff  unless it's super clear what's happening. A supplier<biconsumer makes me go crazy

TUsama replied:
I move the strategy selection into the IParticleSpawnMaterial, I guess this is better than a horrible supplier of BiConsumer

RobertSkalko commented on src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/GLUtils.java:
```diff
@@ -0,0 +1,14 @@
+
+public class GLUtils {
+
+    public static void renderAlwaysSeenText(Runnable render){
```
What's the purpose of this? I see the name but were some particles not displaying? 
It seems like it could cause trouble with the render calls

TUsama replied:
health bar will cover the particle.

RobertSkalko commented on src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/InteractionNotifier.java:
```diff
@@ -9,6 +10,7 @@
 public class InteractionNotifier {
 
     public static void notifyClient(IParticleSpawnMaterial notifier, ServerPlayer source, LivingEntity target){
+        if (!ClientConfigs.getConfig().ENABLE_FLOATING_DMG.get()) return;
```
What did i say about not using client code on the server

TUsama committed:
move config to client side

TUsama committed:
move the strategy selection into the IParticleSpawnMaterial

RobertSkalko commented:
Did you test on server this time

TUsama commented:
yeah, it works this time, no NPE or mis-invoke

RobertSkalko commented on src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/InteractionResultHandler.java:
```diff
@@ -0,0 +1,118 @@
+                }
+            }
+        },
+                (type, entity) -> {
```
Why isn't damageStrategy etc just a field in ParticleSpawnType  ? Seems weird an enum needs to be able to react to any variation of another enum

RobertSkalko replied:
I'd rename ClientReactionStrategy  to ParticlePositionPicker and ParticleSpawnType  to ExileParticleType

TUsama replied:
I don't get you, you mean to break down the ClientReactionStrategy?

RobertSkalko replied:
Yeah it should just tell ParticleSpawnType  the pos and other info to spawn the particle and ParticleSpawnType should be spawning the particle.

TUsama replied:
The ClientReactionStrategy is not just a position picker. See the damageStrategy, the DEFAULT will generate multiple particles while the IN_A_ROW will generate only one. 

I also don't wanna merge these two enums, it will make the ParticleSpawnType very chunky and I can't imagine the way I can use config to easily switch the style after merging.

TUsama committed:
make dodge and resist translatable

RobertSkalko commented on src/main/java/com/robertx22/mine_and_slash/vanilla_mc/packets/interaction/IParticleSpawnMaterial.java:
```diff
@@ -25,7 +29,7 @@ public interface IParticleSpawnMaterial {
     void spawnOnClient(Entity entity);
 
-    enum Type implements IParticleSpawnMaterial {
+    enum Type implements IParticleSpawnMaterial, IAutoLocName {
```
If there's only like 2 or 3 things i tend to just use Words enum for it

TUsama replied:
Honestly, I don't like to put things into Words unless they are really trivial or hard to categorize. But I'll do that.

TUsama committed:
rename, and move to Words

RobertSkalko commented on src/main/java/com/robertx22/mine_and_slash/a_libraries/dmg_number_particle/particle/style/Row.java:
```diff
@@ -0,0 +1,122 @@
+    @Override
+    public void renderDamage(ExileInteractionResultParticle particle, VertexConsumer vertexConsumer, Camera camera, float partialTick, PoseStack posestack, String text, int color) {
+        MultiBufferSource.BufferSource multibuffersource$buffersource = Minecraft.getInstance().renderBuffers().bufferSource();
+        boolean isCrit = text.contains("!");
```
I'll probably merge it but this here is the sole reason i'd completely rework the entire thing.
Why not just pass the particle data so you don't have to do regex magic to figure out how to render it?

TUsama committed:
add IParticleRenderMaterial to adjust the current method

TUsama committed:
rename GLUtils to RenderUtils, add two methods

TUsama committed:
fix format issue on DEFAULT mode

RobertSkalko commented:
Tell me if the refactor is over

TUsama committed:
Update Default.java

TUsama committed:
Update HealParticle.java

TUsama commented:
I moved all code related to rendering into the particle render method, now the ClientSpawnStrategy handles the function of position picking, particle generation, and conversion from IParticleSpawnMaterial to IParticleRenderMaterial. Noted that I don't think directly passing the IParticleSpawnMaterial to Particle is a more effective way here due to the code structure, so I added IParticleRenderMaterial to avoid passing String while keeping the compatibility.

I've tested it on the server and no crashing.

RobertSkalko commented:
Ok i'll merge to test it in a bit

RobertSkalko commented:
Seems to work
